// reveal.js-symbol-per-slide-progress/symbol-per-slide-progress.js
/**
 * reveal.js symbol per slide progress plugin
 *
 * A plugin which shows every slide as single symbol and the symbol of the current active slide in a different color.
 *
 * Demo https://naamor.github.io/reveal.js-symbol-per-slide-progress/
 *
 * MIT License
 * Copyright (c) 2018 Roman Stocker
 */

const VALID_POSITIONS = ["left", "right", "center"];
const VALID_ALIGNS = ["vertical", "horizontal"];

const RevealSymbolPerSlideProgress = window.SymbolPerSlideProgress || (function () {
    
    // The reveal.js instance this plugin is attached to
    let deck;
    let position;
    let align;
    let symbolColor;
    let listItemActiveColor;

    function initialize() {
        var reveal = document.querySelector(".reveal");

        var div = document.createElement("div");
        div.className = "navigation " + position;

        var ul = document.createElement("ul");
        ul.className = "navigation-list";

        updateNavigation(ul);

        deck.addEventListener("slidechanged", function () {
            updateNavigation(ul);
        });

        div.appendChild(ul);
        reveal.appendChild(div);
    }

    // Check input if it is valid and return it
    function getStringOption(option, VALID_ELEMENTS) {
        if (typeof option === "string" && VALID_ELEMENTS.includes(option.toLowerCase())) {
            return option.toLowerCase();
        }
    }

    // Get a color of the theme
    function getColor(selectorValue) {
        var selector = document.querySelector(selectorValue),
            style = window.getComputedStyle(selector),
            color = style.getPropertyValue("color");
        return color;
    }

    // Update colors
    function updateColors(options = {}) {
        symbolColor = options.symbolColor || getColor(".reveal");
        listItemActiveColor = options.symbolActiveColor || getColor(".controls");
    }

    function updateNavigation(ulist) {
        var ul = ulist || document.querySelector(".navigation-list")
        var totalSlides = deck.getTotalSlides();

        // Reset all child element
        while (ul.firstChild) {
            ul.removeChild(ul.firstChild);
        }

        for (let counter = 1; counter <= totalSlides; counter++) {
            var li = document.createElement("li");
            li.className = getClassName();

            if (deck.getSlidePastCount() + 1 === counter) {
                li.classList.add("active");
                li.style.color = listItemActiveColor;
            } else {
                li.style.color = symbolColor;
            }

            ul.appendChild(li);
        }
    };

    function getClassName() {
        if (align === "horizontal") {
            switch (position) {
                case "right":
                    return "navigation-list-element horizontal-right";
                case "center":
                    return "navigation-list-element horizontal-center";
                case "left":
                default:
                    return "navigation-list-element horizontal-left";
            }
        }

        return "navigation-list-element";
    }

    // Modified from math plugin
    function loadResource(url, type, callback) {
        var head = document.querySelector("head");
        var resource;

        if (type === "script") {
            resource = document.createElement("script");
            resource.type = "text/javascript";
            resource.src = url;
        } else if (type === "stylesheet") {
            resource = document.createElement("link");
            resource.rel = "stylesheet";
            resource.href = url;
        }

        // Wrapper for callback to make sure it only fires once
        var finish = function () {
            if (typeof callback === "function") {
                callback.call();
                callback = null;
            }
        };

        resource.onload = finish;

        // IE
        resource.onreadystatechange = function () {
            if (this.readyState === "loaded") {
                finish();
            }
        };

        // Normal browsers
        head.appendChild(resource);
    }

    function scriptPath() {
        // obtain plugin path from the script element
        var path;
        if (document.currentScript) {
            path = document.currentScript.src.slice(0, -17);
        } else {
            var sel = document.querySelector('script[src$="symbol-per-slide-progress.js"]');
            if (sel) {
                path = sel.src.slice(0, -17);
            }
        }

        return path;
    }

    return {
        id: "symbol-per-slide-progress",

        init: function (reveal) {
            deck = reveal;

            // var config = reveal.getConfig().symbolPerSlideProgress || {};
            // position = getStringOption(config.position, VALID_POSITIONS) || "left";
            // align = getStringOption(config.align, VALID_ALIGNS) || "vertical";
            // symbolColor = config.symbolColor || getColor(".reveal .slides section .slide-number");
            // listItemActiveColor = config.listItemActiveColor || getColor(".reveal .slides section .slide-number");

            // loadResource(scriptPath() + "symbol-per-slide-progress.css", "stylesheet", function () {
            //     initialize();
            // });

             // Set all option defaults
            var options = deck.getConfig().symbolperslideprogress || {};
            var path = options.path || "plugin/symbol-per-slide-progress/";
            if (!path.endsWith('/')) {
                path += '/';
            }

            position = getStringOption(options.position, VALID_POSITIONS) || "left";
            align = getStringOption(options.align, VALID_ALIGNS) || "vertical";
            updateColors(options);

            loadResource(path + "symbol-per-slide-progress.css", "stylesheet");

            initialize();
        },

        updateColors: updateColors,
        updateNavigation: updateNavigation
    }
})();

export default RevealSymbolPerSlideProgress;
